pkg/agent/memberlist/cluster.go

```diff
@@ -0,0 +1,321 @@
+	if err != nil {
+		klog.Errorf("error when listing Nodes: %v", err)
+		return
+	}
+	klog.V(3).Infof("List %d nodes", len(nodes))
+
+	for _, node := range nodes {
+		klog.V(4).Infof("Node %s: %#v", node.Name, node.Status.Addresses)
+		nodeAddr, err := k8s.GetNodeAddr(node)
+		if err != nil {
+			klog.Errorf("Failed to obtain local IP address from K8s node: %w", err)
+			continue
+		}
+		member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+		clusterNodes = append(clusterNodes, member)
+	}
+	return
+}
+
+func (c *Cluster) shouldJoinCluster(nodeName string) bool {
+	ipPools, _ := c.ipPoolInformer.Lister().List(labels.Everything())
+	for _, ip := range ipPools {
+		nodeSel := ip.Spec.NodeSelector
+		if nodeSel == nil {
+			return true
+		}
+		nodeSelect, _ := metav1.LabelSelectorAsSelector(nodeSel)
+		nodes, _ := c.nodeInformer.Lister().List(nodeSelect)
+		for _, nd := range nodes {
+			if nd.Name == nodeName {
+				return true
+			}
+		}
+	}
+
+	return false
+}
+
+func (c *Cluster) addNodeMemberHandler(obj interface{}) {
+	node, ok := obj.(*corev1.Node)
+	if !ok {
+		klog.Errorf("Add node callback error, unexpected object type: %v", obj)
+		return
+	}
+	if c.shouldJoinCluster(node.Name) {
+		c.addMember(node)
+	}
+}
+
+func (c *Cluster) memberNum() int {
+	c.memberRWLock.RLock()
+	defer c.memberRWLock.RUnlock()
+	return len(c.existingMembers)
+}
+
+func (c *Cluster) addMember(node *corev1.Node) {
+	c.memberRWLock.Lock()
+	defer c.memberRWLock.Unlock()
+	nodeAddr, err := k8s.GetNodeAddr(node)
+	if err != nil {
+		klog.Errorf("Failed to obtain local IP address from K8s node: %w", err)
+		return
+	}
+	member := fmt.Sprintf("%s:%d", nodeAddr, c.bindPort)
+	c.existingMembers = append(c.existingMembers, member)
+	c.joinMembers(c.existingMembers)
+}
+
+func (c *Cluster) joinMembers(clusterNodes []string) {
+	n, err := c.memberList.Join(clusterNodes)
+	if err != nil {
+		klog.Errorf("Failed to join cluster: %v, cluster nodes: %#v", err, clusterNodes)
+		return
+	}
+	klog.V(2).Infof("Join cluster: %v, cluster nodes: %+v", n, clusterNodes)
+}
+
+func (c *Cluster) Run(stopCh <-chan struct{}) {
+	if !cache.WaitForNamedCacheSync("MemberListCluster", stopCh, c.ipPoolInformerHasSynced) {
+		return
+	}
+
+	newClusterMembers := c.convertListNodesToMemberlist()
+	expectNodeNum := len(newClusterMembers)
+
+	actualMemberNum := c.memberList.NumMembers()
+	klog.V(3).Infof("Nodes num: %d, actual member num: %d", expectNodeNum, actualMemberNum)
+	if actualMemberNum < expectNodeNum {
+		c.joinMembers(newClusterMembers)
+	}
+
+	c.askClusterMemberHealthz()
+
+	// Memberlist will maintain membership information in the background.
+	timeTicker := time.NewTicker(c.defaultClusterHealthzAskInterval)
+	defer func() {
+		timeTicker.Stop()
+		close(c.nodeEventsCh)
+	}()
+	for {
+		select {
+		case <-stopCh:
+			return
+		case nodeEvent := <-c.nodeEventsCh:
+			c.updateNodeConsistenHash(&nodeEvent)
+		case <-timeTicker.C:
+			c.askClusterMemberHealthz()
```

Comment:
Why it needs to log all members periodically? I think log member update events is enough, otherwise duplicate information will overwhelm log files, evicting useful logs.

Reply:
done